src/apps/txtviewer/txtview.cpp

```diff
@@ -0,0 +1,417 @@
+    setCanvasOptions(lastCanvas);
+
+    doffs.clear();
+
+    for (auto noff : noffs) {
+        char* pLineStart = noff;
+
+        // skip if noff points past the block
+        if (pLineStart >= tBlock + tLen) continue;
+
+        char* pLineEnd = uforward(pLineStart); // first letter
+        if (pLineEnd > tBlock + tLen) pLineEnd = tBlock + tLen; // clamp
+
+        while (pLineEnd < tBlock + tLen) {
+            char backupChar = *pLineEnd;
+            *pLineEnd = '\0';
+
+            // reached newline
+            if (*(pLineEnd - 1) == '\n') {
+                if (maxoffset == -1 || maxoffset <= OFF2ROFF(pLineStart)) doffs.push_back(pLineStart);
+                // TXT_DBG lilka::serial.log("Adding doff at %p\n", doffs.back());
+                *pLineEnd = backupChar;
+                break;
+            }
+
+            // check if line fits
+            if (!isLineWithinCanvas(pLineStart, lastCanvas)) {
+                if (maxoffset == -1 || maxoffset <= OFF2ROFF(pLineStart)) doffs.push_back(pLineStart);
+                // TXT_DBG lilka::serial.log("Adding doff at %p\n", doffs.back());
+                *pLineEnd = backupChar;
+
+                // move backward safely
+                char* prev = ubackward(pLineEnd);
+                if (prev == pLineEnd || prev < tBlock) break; // avoid infinite loop
+                pLineEnd = prev;
+
+                pLineStart = pLineEnd;
+                continue;
+            }
+
+            *pLineEnd = backupChar;
+
+            // advance safely
+            char* next = uforward(pLineEnd);
+            if (next <= pLineEnd || next > tBlock + tLen) break;
+            pLineEnd = next;
+        }
+    }
+}
+
+void TxtView::scrollUp() {
+    TXT_DBG LEP;
+    if (!fp || !lastCanvas) return;
+
+    long currentFileOffset = ftell(fp);
+    // Can't go back
+    if (currentFileOffset == 0) return;
+
+    // Do file refresh
+    long prevNLineOffset = flineback(fp, tBuffer, TXT_BUFFER_SIZE);
+    fseek(fp, prevNLineOffset, SEEK_SET);
+    tBlockRefresh();
+
+    // Refresh offs till first doff
+    long maxoffset = ftell(fp); // stick to current file position
+    nOffsRefresh(maxoffset);
+    dOffsRefresh(maxoffset);
+
+    TXT_DBG lilka::serial.log("current offset %ld, prevline %ld", currentFileOffset, prevNLineOffset);
+    // now we've to find doff before saved one
+    bool found = false;
+    const char* nextOffset = doffs[0];
+    for (auto doff : doffs) {
+        if (OFF2ROFF(doff) == currentFileOffset) {
+            found = true;
+            break;
+        }
+        nextOffset = doff;
+    }
+    if (found) fseek(fp, OFF2ROFF(nextOffset), SEEK_SET);
+    else {
+        TXT_DBG lilka::serial.err("Doff not found, check doffs/noffs consistency");
+        return; // looks like we already here
+    }
+    // should be where need at next refresh
+    tBlockRefreshRequired = true; // to be done in update()
+}
+
+void TxtView::scrollDown() {
```

Comment:
Doesn't work correctly.
Check with file 'latin.txt' from attached archive – all file is visible on the screen, but when you press Down button, application scroll the screen buffer up with new blank line. After the last non-blank line is no longer visible, pressing the Down button returns to the beginning of the file.

[test-txts.zip](https://github.com/user-attachments/files/23427868/test-txts.zip)